### PR TITLE
AP_Airspeed: add support ASP5033 airspeed sensor

### PR DESCRIPTION
**Pull request by QioTek**
a new custom product differential pressure I2C sensor from www.qio-tek.com, airspeed range 30kpa.

### CODE REVIEW

File: libraries/AP_Airspeed/AP_Airspeed.cpp [hunk 1]
```diff
@@ -83,7 +84,7 @@ const AP_Param::GroupInfo AP_Airspeed::var_info[] = {
     // @Param: _TYPE
     // @DisplayName: Airspeed type
     // @Description: Type of airspeed sensor
-    // @Values: 0:None,1:I2C-MS4525D0,2:Analog,3:I2C-MS5525,4:I2C-MS5525 (0x76),5:I2C-MS5525 (0x77),6:I2C-SDP3X,7:I2C-DLVR-5in,8:UAVCAN,9:I2C-DLVR-10in,10:I2C-DLVR-20in,11:I2C-DLVR-30in,12:I2C-DLVR-60in,13:NMEA water speed,14:MSP
+    // @Values: 0:None,1:I2C-MS4525D0,2:Analog,3:I2C-MS5525,4:I2C-MS5525 (0x76),5:I2C-MS5525 (0x77),6:I2C-SDP3X,7:I2C-DLVR-5in,8:UAVCAN,9:I2C-DLVR-10in,10:I2C-DLVR-20in,11:I2C-DLVR-30in,12:I2C-DLVR-60in,13:NMEA water speed,14:MSP，15:I2C-ASP5033
```
Comment (Hwurzburg) [inline libraries/AP_Airspeed/AP_Airspeed.cpp:87]: I dont think spaces are allowed in the metadata values field...I might be wrong, though

Comment (QioTek) [reply libraries/AP_Airspeed/AP_Airspeed.cpp]: My fault, I used SBC case "，". @Hwurzburg

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        _dev->read_registers(Press_CSB, &pressraw_bytes[1], 1);
+        _dev->read_registers(Press_LSB, &pressraw_bytes[2], 1); 
+        pressraw = (pressraw_bytes[0]<<16) + (pressraw_bytes[1]<<8)+ pressraw_bytes[2];
+        //Read ADC output data of temperature
+        _dev->read_registers(Temp_MSB, &tempraw_bytes[0], 1);
+        _dev->read_registers(Temp_LSB, &tempraw_bytes[1], 1);
+        tempraw = (tempraw_bytes[0]<<8 )+tempraw_bytes[1];
+        //Judge pressure data direction
+        if (pressraw> Press_neg){
+        pressraw = (pressraw - Press_Rng);
```
Comment (amilcarlucas) [inline libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp:94]: Indent

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        pressure_sum += pressraw / Press_RngCal; 
+        }
+        else{
+        pressure_sum += pressraw / Press_RngCal;
```
Comment (amilcarlucas) [inline libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp:98]: Indent

**Review by amilcarlucas — CHANGES_REQUESTED** [review]
Maybe you can run your newly added file thru Ardupilot Astyle's?
And please squash this into a single commit

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+{
+}
+
+// probe and initialise the sensor
+bool AP_Airspeed_ASP5033::init()
+{
+    _dev = hal.i2c_mgr->get_device(get_bus(), ASP5033_I2C_ADDR);
+    if (!_dev) {
+        return false;
+        printf("ASP5033: sensor no found\n");
```
Comment (tridge) [inline libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp:57]: use GCS_SEND_TEXT() to GCS can see this

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    }
+    _dev->get_semaphore()->take_blocking();
+    _dev->set_retries(2);
+    _dev->get_semaphore()->give();
+    _dev->register_periodic_callback(1000000UL/80U,
+                                    FUNCTOR_BIND_MEMBER(&AP_Airspeed_ASP5033::timer, void));
+    printf("ASP5033: sensor found\n");
```
Comment (tridge) [inline libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp:64]: GCS_SEND_TEXT here too

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+// read the values from the sensor
+void AP_Airspeed_ASP5033::timer()
+{
+    uint8_t SysID;
+    uint8_t Sco;
+    uint8_t pressraw_bytes[3];
+    uint8_t tempraw_bytes[2];
+    _dev->read_registers(Sys_config, &SysID, 1); //Read Sensor_config
+     SysID = (SysID&0xFD);
+    _dev->write_register(Sys_config, SysID); //Set Sensor_config
```
Comment (tridge) [inline libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp:77]: do you need to set config on every loop? can you do this once in init?

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    _dev->write_register(CMD, CMD_measure); //once temperature conversion immediately followed by once sensor signal conversion
+    _dev->read_registers(CMD, &Sco, 1); //Judge whether Data collection is over
+       while ((Sco & 0x08) > 0){
```
Comment (tridge) [inline libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp:80]: don't loop forever, if the cable comes loose we'd lose all devices on this bus

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    _dev->read_registers(Sys_config, &SysID, 1); //Read Sensor_config
+     SysID = (SysID&0xFD);
+    _dev->write_register(Sys_config, SysID); //Set Sensor_config
+    _dev->write_register(CMD, CMD_measure); //once temperature conversion immediately followed by once sensor signal conversion
+    _dev->read_registers(CMD, &Sco, 1); //Judge whether Data collection is over
+       while ((Sco & 0x08) > 0){ 
+       _dev->read_registers(CMD, &Sco, 1);
+       }
+       //Read ADC output data of pressure
+        _dev->read_registers(Press_MSB, &pressraw_bytes[0], 1);
```
Comment (tridge) [inline libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp:84]: can you do a single read of multiple registers?

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp (hunk 2 above)
Comment (tridge) [inline libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp:94]: don't do the sign handling in the driver, instead let the AP_Airspeed code handle it or let user use the tube order parameters

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        //Judge pressure data direction
+        if (pressraw> Press_neg){
+        pressraw = (pressraw - Press_Rng);
+        pressure_sum += pressraw / Press_RngCal; 
+        }
+        else{
+        pressure_sum += pressraw / Press_RngCal;
+        }
+        //Temperature positive and negative processing
+        if(tempraw >Temp_neg){
```
Comment (tridge) [inline libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp:101]: this looks really odd, are you sure it's right?

**Review by tridge — CHANGES_REQUESTED** [review]
please also squash commits

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.h [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+/*
+  backend driver for airspeed from I2C
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Param/AP_Param.h>
+#include <AP_HAL/utility/OwnPtr.h>
+#include <AP_HAL/I2CDevice.h>
+#include <utility>
```
Comment (peterbarker) [inline libraries/AP_Airspeed/AP_Airspeed_ASP5033.h:25]: Why is this include needed?

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.h [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+/*
+  backend driver for airspeed from I2C
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Param/AP_Param.h>
```
Comment (peterbarker) [inline libraries/AP_Airspeed/AP_Airspeed_ASP5033.h:22]: Why is this include needed?

**Comment by WickedShell** [comment]
Is the sensor going to be publicly available? I tried looking at the linked website but it appears to be down.

File: libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp (hunk 4 above)
Comment (QioTek) [reply libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp:57]: I got it, will modify them soon.

**Comment by QioTek** [comment]
@WickedShell  Yep, sensor will go public on sale. It's going to be an economic solution. The website we are currently performing some scheduled maintenance, will be back as soon as possible.

**Pull request closed without merging**